lib/get-eslint-modules.js

```diff
@@ -0,0 +1,44 @@
+var Module = require("module");
+var path = require("path");
+
+/**
+ * We can't just require in libraries related to ESLint, scope traversal etc.
+ * We need to look them up based on what the end user has installed.
+ *
+ * This function handles tracking down all the relevant libraries in this area
+ * and returns an object which includes them as properties.
+ */
+module.exports = function getESLintModules() {
+  try {
+    // avoid importing a local copy of eslint, try to find a peer dependency
+    var eslintLoc = Module._resolveFilename("eslint", module.parent);
+  } catch (err) {
+    try {
+      // avoids breaking in jest where module.parent is undefined
+      eslintLoc = require.resolve("eslint");
+    } catch (err) {
+      throw new ReferenceError("couldn't resolve eslint");
+    }
+  }
+
+  // get modules relative to what eslint will load
+  var eslintMod = new Module(eslintLoc);
+  eslintMod.filename = eslintLoc;
+  eslintMod.paths = Module._nodeModulePaths(path.dirname(eslintLoc));
+
+  var eslintScope = eslintMod.require("eslint-scope");
+  var Definition = eslintMod.require("eslint-scope/lib/definition").Definition;
+  var Referencer = eslintMod.require("eslint-scope/lib/referencer");
+  var Traverser = eslintMod.require("eslint/lib/util/traverser");
+  var ScopeManager = eslintMod.require("eslint-scope/lib/scope-manager");
```

Comment:
- I think `babel-eslint` should have `eslint-scope` in own `dependencies` of package.json. It can pin the version of `eslint-scope` until it gets public API to override its behavior.
- I think `babel-eslint` should have own visitor keys instead of it extends ESLint's.
- I think about `fallback: Traverser.getKeys`, we can move the `fallback` logic from ESLint to `eslint-scope`, so it's shared with all parsers.